### PR TITLE
feat: Allow lists in py expressions

### LINKED ISSUE #114: Support lists in `py()` expressions

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #114

### CODE REVIEW

File: tests/integration/test_py.py [hunk 1]
```diff
@@ -60,3 +60,36 @@ def foo() -> int:
         return x
 
     validate(foo)
+
+
+def test_list_basic(validate):
+    @guppy
+    def foo() -> list[int]:
+        xs = py([1, 2, 3])
```
Comment (croyzor) [inline tests/integration/test_py.py:68]: It's nice that this doesn't require a type annotation

File: guppylang/prelude/_internal.py [hunk 2]
```diff
@@ -67,6 +74,11 @@ def float_value(f: float) -> val.Value:
     return val.ExtensionVal(c=(ConstF64(value=f),))
 
 
+def list_value(v: list[val.Value]) -> val.Value:
+    """Returns the Hugr representation of a list value."""
+    return val.ExtensionVal(c=(ListValue(value=v),))
```
Comment (croyzor) [inline guppylang/prelude/_internal.py:79]: What does the `,` do here?

File: guppylang/checker/expr_checker.py [hunk 3]
```diff
@@ -497,34 +512,7 @@ def visit_ListComp(self, node: ast.ListComp) -> tuple[ast.expr, GuppyType]:
...
-            # Remove the top frame pointing to the `eval` call from the stack trace
-            tb = e.__traceback__.tb_next if e.__traceback__ else None
-            raise GuppyError(
-                "Error occurred while evaluating Python expression:\n\n"
-                + "".join(traceback.format_exception(type(e), e, tb)),
-                node,
-            ) from e
-
+        python_val = eval_py_expr(node, self.ctx)
         if ty := python_value_to_guppy_type(python_val, node, self.ctx.globals):
```
[the remark this answers is not in the document]
Comment (croyzor) [reply guppylang/checker/expr_checker.py:516]: Okay, I'm looking up how python `match` works...

File: tests/error/py_errors/list_different_tys.err [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Guppy compilation failed. Error in file $FILE:6
+
+4:    @guppy
+5:    def foo() -> int:
+6:        return py([1, 1.0])
+                 ^^^^^^^^^^^^
+GuppyError: Python expression of type `<class 'list'>` is not supported by Guppy
```
[the remark this answers is not in the document]
Comment (croyzor) [reply tests/error/py_errors/list_different_tys.err:7]: Or is it that `float` isn't supported by guppy?

[the remark this answers is not in the document]
Comment (mark-koch) [reply tests/error/py_errors/list_different_tys.err]: > This error doesn't seem right... It should be a type mismatch

Right, I improved the error message

File: guppylang/checker/expr_checker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mark-koch) [reply guppylang/checker/expr_checker.py:517]: Here, we're checking that the Python expression inside `py(...)` evaluates to something that is valid in Guppy and compute the corresponding Guppy type

File: guppylang/prelude/_internal.py (hunk 2 above)
Comment (mark-koch) [reply guppylang/prelude/_internal.py:79]: `(x,)` is a unary tuple. We need this because of some serialisation shenanigans...

**Review by croyzor — APPROVED** [review]

**Pull request merged**